src/gwt/src/org/rstudio/core/client/virtualscroller/virtualscroller.js

```diff
@@ -0,0 +1,299 @@
+/*
+ * virtualscroller.js
+ *
+ * Copyright (C) 2020 by RStudio, PBC
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+var VirtualScroller;
+
+(function () {
+  "use strict";
+  VirtualScroller = function () {
+  };
+
+  VirtualScroller.prototype = {
+    setup: function(element) {
+      if (!element)
+        return;
+
+      //  *** _CONSTANTS ***
+      this._DEBUG = true;
+      this._SCROLL_DEBOUNCE_MS = 500;
+      this._BUCKET_MAX_SIZE = 50;
+      this._MAX_VISIBLE_BUCKETS = 10;
```

Comment:
These parameters are probably somewhat machine dependent (i.e. with more hardware you can have more buckets w/o noticeable performance degradation), so it'd be nice if they could be set via an advanced (no UI) option; your call.

Reply:
I partially also picked this due to it just feeling better, in my opinion, to have a smaller scrollbar and having more fine motor control over what you see in the window. Once the window becomes saturated it's quite difficult to quickly get to the line that you want, it's why I `clear` in my terminal pretty often. 

I'll file an enhancement to make this customizable though.

Reply:
There's also interaction with our server-side (I think) console-output limiting logic to consider. For example, try this sequence:

```R
> for (i in 1:2000) cat(c(i, "\n")

1000
1001
...
2000

Notice that you only see the output from 1000 - 2000 (but still see the original prompt with no indication stuff is missing in between; prior to this change the prompt would have been lost, too).

```R
> for (i in 2001:4000) cat(c(i, "\n"))
3000
3001
...
4000
```

Now you can scroll back and indeed are seeing > 1000 lines, but these gaps are unexpected, I think.